File: game/forest/talking_after_hurt.py
import pygame

from animation import Animation
from audio_helper import AudioHelper
from forest.forest_resources import ForestResources
from forest.playing import Playing
from game_data import GameData
from render_type import RenderType
from state import State


class TalkingAfterHurt(State):
    def __init__(self, context: GameData):
        super().__init__(context)
        self.needs_background = RenderType.PRE
        self.needs_bottom = RenderType.PRE

    def process_events(self, events):
        if (self.context.forest_lives == 1 and self.get_frame_index() >= len(ForestResources.sync_lastlife)) or \
            (self.context.forest_lives >= 1 and self.get_frame_index() >= len(ForestResources.sync_dieonce)):
            return Playing

        if self.one_shot(0.5, "knock1"):
            AudioHelper.play(ForestResources.sfx_hugo_knock, self.context.audio_port)

        if self.one_shot(0, "speak"):
            AudioHelper.play(ForestResources.sfx_hugo_knock, self.context.audio_port)

            if self.context.forest_lives == 1:
                AudioHelper.play(ForestResources.speak_lastlife, self.context.audio_port)
            else:
                AudioHelper.play(ForestResources.speak_dieonce, self.context.audio_port)

        return None

    def render(self, screen):
        sync = ForestResources.sync_lastlife if self.context.forest_lives == 1 else ForestResources.sync_dieonce
        screen.blit(Animation.get_sync_frame(ForestResources.hugo_telllives, sync, self.get_frame_index()), (128, -16))

        if self.get_frame_index() < 8 and self.get_frame_index() % 4 == 0:
            screen.blit(ForestResources.hugo_hand2[0], (96, 78))
        else:
            screen.blit(ForestResources.hugo_hand1[0], (96, 83))

File: game-web/src/forest/talkingAfterHurt.ts
/** Port of game/forest/talking_after_hurt.py */

import { State, type StateClass } from '../state';
import { Animation } from '../animation';
import { AudioHelper } from '../audioHelper';
import { ForestResources } from './forestResources';
import { Playing } from './playing';
import { RenderType } from '../renderType';
import type { GameData } from '../gameData';
import type { PhoneEvents } from '../phoneEvents';

export class TalkingAfterHurt extends State {
  readonly needsBackground = RenderType.PRE;
  readonly needsBottom = RenderType.PRE;

  constructor(context: GameData) { super(context); }

  processEvents(_events: PhoneEvents): StateClass | null {
    if (
      (this.context.forestLives === 1 && this.getFrameIndex() >= ForestResources.syncLastlife.length) ||
      (this.context.forestLives >= 1 && this.getFrameIndex() >= ForestResources.syncDieonce.length)
    ) {
      return Playing;
    }

    if (this.oneShot(0.5, 'knock1')) AudioHelper.play(ForestResources.sfxHugoKnock);
    if (this.oneShot(0, 'speak')) {
      AudioHelper.play(ForestResources.sfxHugoKnock);
      if (this.context.forestLives === 1) {
        AudioHelper.play(ForestResources.speakLastlife);
      } else {
        AudioHelper.play(ForestResources.speakDieonce);
      }
    }

    return null;
  }

  render(ctx: OffscreenCanvasRenderingContext2D): void {
    const sync = this.context.forestLives === 1 ? ForestResources.syncLastlife : ForestResources.syncDieonce;
    ctx.drawImage(Animation.getSyncFrame(ForestResources.hugoTelllives, sync, this.getFrameIndex()), 128, -16);

    if (this.getFrameIndex() < 8 && this.getFrameIndex() % 4 === 0) {
      ctx.drawImage(ForestResources.hugoHand2[0], 96, 78);
    } else {
      ctx.drawImage(ForestResources.hugoHand1[0], 96, 83);
    }
  }
}
